src/lib/message_filters.py

```diff
@@ -30,3 +31,24 @@ def filter_time_range(messages, time_key, start_time, end_time):
     @staticmethod
     def filter_noise(messages, message_key, noise_fn):
         return [td for td in messages if not noise_fn(td.get(message_key))]
+    
+    @staticmethod
+    def sub_sample_messages(messages):
+        uuids =[]
+        hashed_uuids  = []
+        sub_sample_hashed_uuids = []
+        sub_sample_data = []
+        
+        for td in messages:
+            if td["avf_phone_id"] not in uuids:
+                uuids.append(td["avf_phone_id"])
+                hashed_uuids.append(int(hashlib.sha256(td["avf_phone_id"].encode('utf-8')).hexdigest(), 16))
+
+                for hashed_uuid in hashed_uuids:
+                    if int(str(hashed_uuid)[0]) < 4:
+                        sub_sample_hashed_uuids.append(hashed_uuid)
+            
+            for hashed_uuid in sub_sample_hashed_uuids:
```

Comment:
By appending this TracedData for each sub_sample_hashed_uuid you've seen so far, this will append the same td to your sub_sample_data many times. If you logged the number of messages going in/out of this function that would enable you to detect this (e.g. as was done here https://github.com/AfricasVoices/Project-ADSS/pull/53)

Reply:
Also this logic seems unnecessarily complex. Why do you need to store all those uuids, can't you just:

for td in messages:
   if hash(td["uid"]) < threshold: 
        subsample.append(td)